File: FitProject/FitProject/Core/Sync/SyncEngine.swift
import Foundation
import FirebaseFirestore
import Combine

@MainActor
final class SyncEngine: ObservableObject {
    @Published var isSyncing = false
    @Published var lastSyncDate: Date?
    @Published var syncError: String?

    private var listeners: [ListenerRegistration] = []
    private let firestore = FirestoreService.shared

    func startSync(userId: String, appState: AppState) {
        stopSync()

        let logsListener = firestore.listenToWorkoutLogs(userId: userId) { logs in
            Task { @MainActor in
                appState.workoutLogs = logs
                appState.updateWeeklyProgress()
            }
        }
        listeners.append(logsListener)

        let habitsListener = firestore.listenToHabits(userId: userId) { habits in
            Task { @MainActor in
                appState.habits = habits
            }
        }
        listeners.append(habitsListener)
    }

    func stopSync() {
        listeners.forEach { $0.remove() }
        listeners.removeAll()
    }

    func fullSync(userId: String, appState: AppState) async {
        isSyncing = true
        syncError = nil
        defer {
            isSyncing = false
            lastSyncDate = Date()
        }

        do {
            async let assignedPrograms = firestore.fetchAssignedPrograms(userId: userId)
            async let creatorPrograms = firestore.fetchCreatorPrograms(userId: userId)
            async let workoutLogs = firestore.fetchWorkoutLogs(userId: userId)
            async let habits = firestore.fetchHabits(userId: userId)
            async let progressPictures = firestore.fetchProgressPictures(userId: userId)
            async let measurements = firestore.fetchAllMeasurements(userId: userId)
            async let content = firestore.fetchContent(userId: userId)
            async let records = firestore.fetchPersonalRecords(userId: userId)
            async let forms = firestore.fetchForms(userId: userId)
            async let profile = firestore.fetchUserProfile(userId: userId)

            let (assigned, creator, logs, habitsResult, pictures, meas, cont, recs, formsResult, userProfile) = try await (
                assignedPrograms, creatorPrograms, workoutLogs, habits, progressPictures,
                measurements, content, records, forms, profile
            )

            var allPrograms: [FPProgram] = creator
            for item in assigned {
                if let program = item.program, !allPrograms.contains(where: { $0.id == program.id }) {
                    var p = program
                    p.completedWorkouts = item.completedWorkouts
                    allPrograms.append(p)
                }
            }

            var weeksByProgram: [String: [FPProgramWeek]] = [:]
            for program in allPrograms {
                weeksByProgram[program.id] = try await firestore.fetchProgramWeeks(programId: program.id)
            }

            appState.programs = allPrograms
            appState.programWeeks = weeksByProgram
            appState.assignedPrograms = assigned
            appState.workoutLogs = logs
            appState.habits = habitsResult
            appState.allProgressPictures = pictures
            appState.progressSessions = FirestoreService.groupProgressSessions(pictures)
            appState.measurements = meas
            appState.content = cont
            appState.personalRecords = recs
            appState.forms = formsResult
            appState.unitPreferences = userProfile.unitPreferences
            appState.authService.updateUserProfile(userProfile)
            appState.updateWeeklyProgress()
            appState.selectNextWorkout()
        } catch {
            syncError = error.localizedDescription
        }
    }

    func pushWorkoutLog(_ log: FPWorkoutLog) async throws {
        try await firestore.saveWorkoutLog(log)
    }

    func pushHabitUpdate(habit: FPHabit, userId: String, value: Double) async throws {
        try await firestore.saveUserHabitLog(habit, userId: userId, value: value)
    }

    func pushMeasurement(_ measurement: FPMeasurement, userId: String) async throws {
        try await firestore.saveMeasurement(measurement, userId: userId)
    }

    func pushPersonalRecord(_ record: FPPersonalRecord, userId: String) async throws {
        try await firestore.savePersonalRecord(record, userId: userId)
    }

    func pushFormSubmission(formId: String, userId: String, answers: [FPFormAnswer], appState: AppState) async throws {
        try await firestore.submitForm(formId: formId, clientId: userId, answers: answers)
        if let index = appState.forms.firstIndex(where: { $0.id == formId }) {
            var form = appState.forms[index]
            form.submissions.append(FPFormSubmission(
                clientId: userId,
                submittedAt: Date(),
                answers: answers
            ))
            appState.forms[index] = form
        }
    }

    func pushUnitPreference(userId: String, key: String, value: String, appState: AppState) async throws {
        try await firestore.updateUnitPreference(userId: userId, key: key, value: value)
        var prefs = appState.unitPreferences
        switch key {
        case "weight": prefs.weight = value
        case "mass": prefs.mass = value
        case "circumference": prefs.circumference = value
        case "distance": prefs.distance = value
        case "time": prefs.time = value
        default: break
        }
        appState.unitPreferences = prefs
        if var user = appState.authService.currentUser {
            user.unitPreferences = prefs
            appState.authService.updateUserProfile(user)
        }
    }

    func pushProgressPhotoSession(draft: FPProgressPhotoDraft, userId: String, appState: AppState) async throws {
        var poses: [(poseType: String, imageUrl: String, existingId: String?)] = []
        for (poseType, payload) in draft.poseImageData {
            let imageUrl = try await FirebaseStorageService.shared.uploadProgressPhoto(
                userId: userId,
                sessionId: draft.sessionId,
                poseType: poseType,
                data: payload.data,
                contentType: payload.contentType,
                fileExtension: payload.fileExtension
            )
            poses.append((poseType, imageUrl, nil))
        }

        try await firestore.saveProgressSession(
            userId: userId,
            sessionId: draft.sessionId,
            dateCreated: draft.dateCreated,
            notes: draft.notes.isEmpty ? nil : draft.notes,
            poses: poses
        )

        let pictures = poses.map { pose in
            FPProgressPicture(
                id: UUID().uuidString,
                userId: userId,
                sessionId: draft.sessionId,
                poseType: pose.poseType,
                imageUrl: pose.imageUrl,
                dateCreated: draft.dateCreated,
                notes: draft.notes.isEmpty ? nil : draft.notes
            )
        }
        appState.allProgressPictures.insert(contentsOf: pictures, at: 0)
        appState.progressSessions = FirestoreService.groupProgressSessions(appState.allProgressPictures)
    }
}

File: FitProjectDroid/app/src/main/java/com/fitproject/droid/data/SyncEngine.kt
package com.fitproject.droid.data

import com.google.firebase.firestore.ListenerRegistration
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import java.util.Date

data class FullSyncResult(
    val programs: List<FPProgram>,
    val programWeeks: Map<String, List<FPProgramWeek>>,
    val assignedPrograms: List<FPAssignedProgram>,
    val workoutLogs: List<FPWorkoutLog>,
    val habits: List<FPHabit>,
    val allProgressPictures: List<FPProgressPicture>,
    val progressSessions: List<FPProgressSession>,
    val measurements: List<FPMeasurement>,
    val content: List<FPContent>,
    val personalRecords: List<FPPersonalRecord>,
    val forms: List<FPForm>,
    val unitPreferences: FPUnitPreferences,
    val userProfile: FPUser
)

interface SyncStateCallbacks {
    fun onWorkoutLogsUpdated(logs: List<FPWorkoutLog>)
    fun onHabitsUpdated(habits: List<FPHabit>)
    fun onFullSyncResult(result: FullSyncResult)
    fun onSyncingChanged(isSyncing: Boolean)
    fun onLastSyncDateChanged(date: Date?)
    fun onSyncError(error: String?)
    fun onFormsUpdated(forms: List<FPForm>)
    fun onUnitPreferencesUpdated(prefs: FPUnitPreferences, user: FPUser?)
    fun onProgressPicturesUpdated(
        pictures: List<FPProgressPicture>,
        sessions: List<FPProgressSession>
    )
}

class SyncEngine {
    private val firestore = FirestoreService.shared
    private val storage = StorageService.shared
    private val listeners = mutableListOf<ListenerRegistration>()

    var isSyncing: Boolean = false
        private set
    var lastSyncDate: Date? = null
        private set
    var syncError: String? = null
        private set

    fun startSync(userId: String, callbacks: SyncStateCallbacks) {
        stopSync()

        val logsListener = firestore.listenToWorkoutLogs(userId) { logs ->
            callbacks.onWorkoutLogsUpdated(logs)
        }
        listeners.add(logsListener)

        val habitsListener = firestore.listenToHabits(userId) { habits ->
            callbacks.onHabitsUpdated(habits)
        }
        listeners.add(habitsListener)
    }

    fun stopSync() {
        listeners.forEach { it.remove() }
        listeners.clear()
    }

    suspend fun fullSync(userId: String, callbacks: SyncStateCallbacks) {
        isSyncing = true
        syncError = null
        callbacks.onSyncingChanged(true)
        callbacks.onSyncError(null)

        val syncErrors = mutableListOf<String>()
        try {
            coroutineScope {
                val assignedDeferred = async { firestore.fetchAssignedPrograms(userId) }
                val creatorDeferred = async { firestore.fetchCreatorPrograms(userId) }
                val logsDeferred = async { firestore.fetchWorkoutLogs(userId) }
                val habitsDeferred = async { firestore.fetchHabits(userId) }
                val picturesDeferred = async { firestore.fetchProgressPictures(userId) }
                val measurementsDeferred = async { firestore.fetchAllMeasurements(userId) }
                val contentDeferred = async { firestore.fetchContent(userId) }
                val recordsDeferred = async { firestore.fetchPersonalRecords(userId) }
                val formsDeferred = async { firestore.fetchForms(userId) }
                val profileDeferred = async {
                    runCatching { firestore.fetchUserProfile(userId) }
                        .onFailure { syncErrors.add(it.message ?: "Failed to load profile") }
                        .getOrNull()
                }

                val assigned = assignedDeferred.await()
                val creator = creatorDeferred.await()
                val logs = logsDeferred.await()
                val habitsResult = habitsDeferred.await()
                val pictures = picturesDeferred.await()
                val meas = measurementsDeferred.await()
                val cont = contentDeferred.await()
                val recs = recordsDeferred.await()
                val formsResult = formsDeferred.await()
                val userProfile = profileDeferred.await()

                val allPrograms = creator.toMutableList()
                for (item in assigned) {
                    val program = item.program ?: continue
                    if (allPrograms.none { it.id == program.id }) {
                        allPrograms.add(program.copy(completedWorkouts = item.completedWorkouts))
                    }
                }

                val weeksByProgram = mutableMapOf<String, List<FPProgramWeek>>()
                for (program in allPrograms) {
                    weeksByProgram[program.id] = firestore.fetchProgramWeeks(program.id)
                }

                if (allPrograms.isEmpty() && syncErrors.isNotEmpty()) {
                    throw IllegalStateException(syncErrors.first())
                }

                val result = FullSyncResult(
                    programs = allPrograms,
                    programWeeks = weeksByProgram,
                    assignedPrograms = assigned,
                    workoutLogs = logs,
                    habits = habitsResult,
                    allProgressPictures = pictures,
                    progressSessions = firestore.groupProgressSessions(pictures),
                    measurements = meas,
                    content = cont,
                    personalRecords = recs,
                    forms = formsResult,
                    unitPreferences = userProfile?.unitPreferences ?: FPUnitPreferences(),
                    userProfile = userProfile ?: FPUser(id = userId)
                )
                callbacks.onFullSyncResult(result)
                syncError = null
                callbacks.onSyncError(null)
            }
        } catch (e: Exception) {
            syncError = e.message
            callbacks.onSyncError(e.message)
        } finally {
            isSyncing = false
            lastSyncDate = Date()
            callbacks.onSyncingChanged(false)
            callbacks.onLastSyncDateChanged(lastSyncDate)
        }
    }

    suspend fun pushWorkoutLog(log: FPWorkoutLog) {
        firestore.saveWorkoutLog(log)
    }

    suspend fun pushHabitUpdate(habit: FPHabit, userId: String, value: Double) {
        firestore.saveUserHabitLog(habit, userId, value)
    }

    suspend fun pushMeasurement(measurement: FPMeasurement, userId: String) {
        firestore.saveMeasurement(measurement, userId)
    }

    suspend fun pushPersonalRecord(record: FPPersonalRecord, userId: String) {
        firestore.savePersonalRecord(record, userId)
    }

    suspend fun pushFormSubmission(
        formId: String,
        userId: String,
        answers: List<FPFormAnswer>,
        callbacks: SyncStateCallbacks,
        currentForms: List<FPForm>
    ) {
        firestore.submitForm(formId, userId, answers)
        val updated = currentForms.map { form ->
            if (form.id == formId) {
                form.copy(
                    submissions = form.submissions + FPFormSubmission(
                        clientId = userId,
                        submittedAt = Date(),
                        answers = answers
                    )
                )
            } else form
        }
        callbacks.onFormsUpdated(updated)
    }

    suspend fun pushUnitPreference(
        userId: String,
        key: String,
        value: String,
        callbacks: SyncStateCallbacks,
        currentPrefs: FPUnitPreferences,
        currentUser: FPUser?
    ) {
        firestore.updateUnitPreference(userId, key, value)
        val prefs = when (key) {
            "weight" -> currentPrefs.copy(weight = value)
            "mass" -> currentPrefs.copy(mass = value)
            "circumference" -> currentPrefs.copy(circumference = value)
            "distance" -> currentPrefs.copy(distance = value)
            "time" -> currentPrefs.copy(time = value)
            else -> currentPrefs
        }
        val user = currentUser?.copy(unitPreferences = prefs)
        callbacks.onUnitPreferencesUpdated(prefs, user)
    }

    suspend fun pushProgressPhotoSession(
        draft: FPProgressPhotoDraft,
        userId: String,
        callbacks: SyncStateCallbacks,
        currentPictures: List<FPProgressPicture>
    ) {
        val poses = mutableListOf<Triple<String, String, String?>>()
        for ((poseType, payload) in draft.poseImageData) {
            val imageUrl = storage.uploadProgressPhoto(
                userId = userId,
                sessionId = draft.sessionId,
                poseType = poseType,
                data = payload.data,
                contentType = payload.contentType,
                fileExtension = payload.fileExtension
            )
            poses.add(Triple(poseType, imageUrl, null))
        }

        firestore.saveProgressSession(
            userId = userId,
            sessionId = draft.sessionId,
            dateCreated = draft.dateCreated,
            notes = draft.notes.ifEmpty { null },
            poses = poses
        )

        val pictures = poses.map { (poseType, imageUrl, _) ->
            FPProgressPicture(
                id = java.util.UUID.randomUUID().toString(),
                userId = userId,
                sessionId = draft.sessionId,
                poseType = poseType,
                imageUrl = imageUrl,
                dateCreated = draft.dateCreated,
                notes = draft.notes.ifEmpty { null }
            )
        }
        val allPictures = pictures + currentPictures
        callbacks.onProgressPicturesUpdated(
            allPictures,
            firestore.groupProgressSessions(allPictures)
        )
    }
}